wfe2/wfe.go

```diff
@@ -842,17 +842,31 @@ func (wfe *WebFrontEndImpl) NewAccount(
 
 	// TODO(#5545): Spending and Refunding can be async until these rate limits
 	// are authoritative. This saves us from adding latency to each request.
-	go wfe.checkNewAccountLimits(ctx, ip)
+	doneCheckingLimits := make(chan struct{})
 	var newRegistrationSuccessful bool
+	var errIsRateLimit bool
+
+	go func() {
+		// Close the channel on goroutine completion.
+		defer close(doneCheckingLimits)
+		wfe.checkNewAccountLimits(ctx, ip)
+	}()
+
 	defer func() {
-		if !newRegistrationSuccessful {
+		if !newRegistrationSuccessful && !errIsRateLimit {
+			// Wait for the rate limit check to complete before attempting to refund
+			// the limits. If the check failed, we don't want to refund anything.
+			<-doneCheckingLimits
 			go wfe.refundNewAccountLimits(ctx, ip)
 		}
 	}()
 
 	// Send the registration to the RA via grpc
 	acctPB, err := wfe.ra.NewRegistration(ctx, &reg)
 	if err != nil {
+		if errors.Is(err, berrors.RateLimit) {
+			errIsRateLimit = true
```

Comment:
Given that we had to have a whole discussion about this for me to understand why it's the right thing to do, we probably want to document why we do this here in the code. That will also give us an easier way to remember that this can/should be removed once the new limits become authoritative.